keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -339,24 +351,301 @@ service CSIProxyService {
     // MountISCSILun mounts a remote LUN over iSCSI and returns the OS disk device number.
     rpc MountISCSILun(MountISCSILunRequest) returns (MountISCSILunResponse) {}
 
-    // LinkVolume invokes mklink on the global staging path of a volume linking it to a path within a container
-    rpc LinkVolume(LinkVolumeRequest) returns (LinkVolumeResponse) {}
+    // LinkPath invokes mklink on the global staging path of a volume linking it to a path within a container
+    rpc LinkPath(LinkPathRequest) returns (LinkPathResponse) {}
+
+    // ListDiskLocations returns locations <Bus, Target, LUN ID> of all disk devices enumerated by Windows
+    rpc ListDiskLocations(ListDiskLocationsRequest) returns (ListDiskLocationsResponse) {}
 
-    // GetDiskLocations returns locations <Bus, Target, LUN ID> of all disk devices enumerated by Windows
-    rpc GetDiskLocations(GetDiskLocationsRequest) returns (GetDiskLocationsResponse) {}
+    // ListDiskIDs returns all IDs (from IOCTL_STORAGE_QUERY_PROPERTY) of all disk devices enumerated by Windows
+    rpc ListDiskIDs(GetDiskIDsRequest) returns (ListDiskIDsResponse) {}
 
-    // GetDiskIDs returns all IDs (from IOCTL_STORAGE_QUERY_PROPERTY) of all disk devices enumerated by Windows
-    rpc GetDiskIDs(GetDiskIDsRequest) returns (GetDiskIDsResponse) {}
+    // ListDiskVolumeMappings returns a map of all disk devices and volumes GUIDs
+    rpc ListDiskVolumeMappings(ListDiskVolumeMappingsRequest) returns (ListDiskVolumeMappingsResponse) {}
 
     // ResizeVolume performs resizing of the partition and file system for a block based volume
     rpc ResizeVolume(ResizeVolumeRequest) returns (ResizeVolumeResponse) {}
+
+    // DismountVolume gracefully dismounts a volume
+    rpc DismountVolume(DismountVolumeRequest) return (DismountVolumeResponse) {}
+}
+
+// Context of the paths used for path prefix validation
+enum PathContext {
+    // plugin maps to the configured kubernetes-csi-plugins-path path prefix
+    PLUGIN = 0;
+    // container maps to the configured container-runtime-storage-path path prefix
+    CONTAINER = 1;
+}
+
+message MkdirRequest {
+    // The path to create in the host filesystem.
+    // All special characters allowed by Windows in path names will be allowed
+    // except for restrictions noted below. For details, please check:
+    // https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file
+    //
+    // Restrictions:
+    // Needs to be an absolute path under kubernetes-csi-plugins-path
+    // or container-runtime-storage-path based on context
+    // Cannot exist already in host
```

Comment:
if it does exist already, is an error returned? is the error distinguishable as an "already exists" error?

Reply:
Yes. I went ahead and added a Windows OS error code for all responses to make error detection and reaction to them easier.